gnucash/import-export/import-backend.c

```diff
@@ -1156,13 +1110,18 @@ gboolean gnc_import_exists_online_id (Transaction *trans)
     /* Look for an online_id in the first split */
     source_split = xaccTransGetSplit(trans, 0);
     g_assert(source_split);
```

Comment:
Because you've removed the atomicity of inserting splits you need to remove *all* of the asserts, even the ones you didn't write, and make sure that once `inc_account_edit` has been called `commit_all` will be too. Note that `g_return_if_fail`, `g_return_val_if_fail` call `g_log` with `G_LOG_LEVEL_ERROR` and that under some conditions that will call `g_assert` so you need to log the error with `PWARN` and return.

Reply:
I'm not getting this. What's a situation where g_assert(source_split) would trigger, that's caused by what I changed? I don't mind removing all the assert, but I'd like to understand why that's necessary!
Thanks for the thorough review BTW!

Reply:
A transaction with a `NULL` split llist. Shouldn't happen, but that doesn't make it impossible... and even without having raised the account edit level it would hardly be a reason to crash the program, just return NULL and make sure that the calling code knows to not add it to the list and continue. Same for the `g_assert(trans)` at the top of the function.

Why it's necessary is that the SQL backend writes the db as part of processing the commit. You're blocking committing any changes to the accounts table but allowing commits to the transactions and splits tables. It's not protected with an encompassing SQL transaction that would cause the SQL server to roll everything back if GnuCash crashes, so a crash would cause a database with reference errors that either wouldn't load or wouldn't balance right and would be difficult to repair.

Even removing all of the asserts can't completely prevent crashes so I'm coming to think that the performance boost isn't worth the risk.

Reply:
Ah yes, I understand now. Transaction changes are allowed but account changes are not, so in case of a crash this would be bad. Here's another option: keep all the commits as they are (i.e. remove all beginEdits) but add code to disable the computation of the balance until the end. In case of a crash, that wouldn't be a big deal at all, but it would still save a bunch of time. 
I might even be simpler to add that code than what I had to add to implement the beginEdit... 
What do you think?

Reply:
That would surely be safer, though I'm skeptical that it would be simpler. What do you have in mind?

Reply:
> That would surely be safer, though I'm skeptical that it would be simpler. What do you have in mind?

How about adding a flag to the account_s structure
`gint defer_bal_computation ;`
that's 0 by default. In the code that I added which calls xaccAccBeingEdit, raise this flag instead (and add the account to the list as is currently done. Add code to not compute the balance when the flag is up. Then the final function that currently calls xaccAccCommitEdit, replace the call by set_defer_bal_computation(false) and in that function trigger the balance computation.
It's annoying to add a flag to the account structure just for that, but it's not a huge deal... It's possible that this flag/function could be reused in other parts of the code when batch modifications have to be made (i,e., move a bunch of transactions, etc).